abn/application/test_helper.go

```diff
@@ -0,0 +1,100 @@
+package application
+
+import (
+	"context"
+	"io/ioutil"
+	"path/filepath"
+	"runtime"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/iter8-tools/iter8/abn/k8sclient"
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func yamlToSecret(folder, file, name string, kClient *k8sclient.KubeClient) error {
+	byteArray, err := readYamlFromFile(folder, file)
+	if err != nil {
+		return err
+	}
+
+	secretName := GetSecretNameFromKey(name)
+	secretNamespace := GetNamespaceFromKey(name)
+
+	_, err = kClient.Typed().CoreV1().Secrets(secretNamespace).Create(context.TODO(), &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: secretNamespace,
+		},
+		StringData: map[string]string{KEY: string(byteArray)},
+	}, metav1.CreateOptions{})
+	return err
+}
+
+func readYamlFromFile(folder, file string) ([]byte, error) {
+	_, filename, _, _ := runtime.Caller(1) // one step up the call stack
+	fname := filepath.Join(filepath.Dir(filename), folder, file)
+
+	return ioutil.ReadFile(fname)
+}
+
+type applicationAssertion struct {
+	namespace, name  string
+	tracks, versions []string
+}
+
+func assertApplication(t *testing.T, a *Application, assertion applicationAssertion) {
+	assert.NotNil(t, a)
+	assert.Contains(t, a.String(), assertion.namespace+"/"+assertion.name)
+
+	assert.Equal(t, assertion.name, GetNameFromKey(a.Name))
+	assert.Equal(t, assertion.namespace, GetNamespaceFromKey(a.Name))
+
+	assert.Len(t, a.Tracks, len(assertion.tracks))
+	for _, track := range assertion.tracks {
+		assert.Contains(t, a.Versions, a.Tracks[track])
+	}
+	assert.Len(t, a.Versions, len(assertion.versions))
+
+	for _, v := range a.Versions {
+		assert.NotNil(t, v.Metrics)
+	}
+}
+
+type versionAssertion struct {
+	track   string
+	metrics []string
+}
+
+func assertVersion(t *testing.T, v *Version, assertion versionAssertion) {
+	assert.NotNil(t, v)
+
+	track := v.GetTrack()
+	if assertion.track == "" {
+		assert.Nil(t, track)
+	} else {
+		assert.Equal(t, assertion.track, *track)
+	}
+
+	assert.Len(t, v.Metrics, len(assertion.metrics))
+	assert.NotNil(t, v.Metrics)
+	for m := range v.Metrics {
+		assert.Contains(t, assertion.metrics, m)
+	}
+}
+
+// Clear the application map
+func (m *ThreadSafeApplicationMap) Clear() {
```

Comment:
Consider an `internal` package for `Clear` and `NumApplications`.

Reply:
An internal package does not seem like the right solution to our challenge. The external functions are accessing package local fields. It is not possible to move them to an internal package without moving the objects that contain those fields. One would need to move pretty much everything and for each item to be externalized a wrapper written to make it public.